### PR TITLE
KAFKA-16614:Disallow @ClusterTemplate("")

### PR DESCRIPTION
**Pull request by TaiJuWu**
As title.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/java/kafka/test/junit/ClusterTestExtensions.java [hunk 1]
```diff
@@ -123,8 +123,7 @@ private void processClusterTemplate(ExtensionContext context, ClusterTemplate an
         if (!annot.value().isEmpty()) {
             generateClusterConfigurations(context, annot.value(), generatedClusterConfigs::add);
         } else {
-            // Ensure we have at least one cluster config
-            generatedClusterConfigs.add(ClusterConfig.defaultClusterBuilder().build());
+            throw new IllegalStateException("Annotation value can't be empty string.");
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: Could we check `annot.value().isEmpty()` first? for example:
```java
if (annot.value().isEmpty()) throw exception;
```
With above change, we can simplify the if-else

**Review by chia7712 — COMMENTED** [review]
@TaiJuWu thanks for this contribution. btw, is this check (https://github.com/apache/kafka/blob/81c222e9779c3339aa139ab930a74aba2c7c8685/core/src/test/java/kafka/test/junit/ClusterTestExtensions.java#L92) unnecessary?

**Comment by TaiJuWu** [comment]
> @TaiJuWu thanks for this contribution. btw, is this check (
> 
> https://github.com/apache/kafka/blob/81c222e9779c3339aa139ab930a74aba2c7c8685/core/src/test/java/kafka/test/junit/ClusterTestExtensions.java#L92
> 
> ) unnecessary?

Yes, I believe it’s unnecessary. If we give default value (code on truncated branch), the size must be greater zero, we just need to select one.

Here we have two solutions, one is current version, the other is you look last time.

I don’t take any position, both are great for me.

**Comment by chia7712** [comment]
> I don’t take any position, both are great for me.

The current solution LGTM, and we can remove unnecessary condition check.

**Comment by TaiJuWu** [comment]
> > I don’t take any position, both are great for me.
> 
> The current solution LGTM, and we can remove unnecessary condition check.

Thanks, the current version is the most simplified.
Just need CI check.

**Comment by chia7712** [comment]
@TaiJuWu Could you add a new test file (`ClusterTestExtensionsUnitTest`) to verify the function `processClusterTemplate`?

File: core/src/test/java/kafka/test/ClusterTestExtensionsUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+public class ClusterTestExtensionsUnitTest {
+    @Test
+    void testProcessClusterTemplate() throws Exception {
+        ClusterTestExtensions ext = mock(ClusterTestExtensions.class);
+        ExtensionContext context = mock(ExtensionContext.class);
+        Consumer<TestTemplateInvocationContext> testInvocations = mock(Consumer.class);
+        ClusterTemplate annot = mock(ClusterTemplate.class);
+        when(annot.value()).thenReturn("");
+
+        Method method = ClusterTestExtensions.class.getDeclaredMethod("processClusterTemplate",
```
Comment (chia7712) [inline core/src/test/java/kafka/test/ClusterTestExtensionsUnitTest.java]: Maybe we can make `processClusterTemplate` be package-private, and then we just test that the method can throw exception when value is empty

Comment (TaiJuWu) [reply core/src/test/java/kafka/test/ClusterTestExtensionsUnitTest.java]: Done. Please review.

**Comment by chia7712** [comment]
@TaiJuWu please fix the conflicts

**Comment by TaiJuWu** [comment]
> @TaiJuWu please fix the conflicts

Done.

File: core/src/test/java/kafka/test/junit/ClusterTestExtensions.java [hunk 3]
```diff
@@ -118,16 +115,15 @@ public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContex
         return generatedContexts.stream();
     }
 
-    private void processClusterTemplate(ExtensionContext context, ClusterTemplate annot,
+    protected void processClusterTemplate(ExtensionContext context, ClusterTemplate annot,
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: please remove `protected` to make it be package-private

File: core/src/test/java/kafka/test/junit/ClusterTestExtensionsUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.TestTemplateInvocationContext;
+import java.lang.reflect.Method;
+import java.util.function.Consumer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class ClusterTestExtensionsUnitTest {
+    @Test
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/ClusterTestExtensionsUnitTest.java:31]: please add `@SuppressWarnings("unchecked")`

**Review by chia7712 — COMMENTED** [review]
@TaiJuWu thanks for this fix. two minor comments left.

File: core/src/test/java/kafka/test/junit/ClusterTestExtensionsUnitTest.java (hunk 4 above)
Comment (TaiJuWu) [reply core/src/test/java/kafka/test/junit/ClusterTestExtensionsUnitTest.java:31]: Done.

File: core/src/test/java/kafka/test/junit/ClusterTestExtensions.java [hunk 5]
```diff
@@ -91,9 +91,6 @@ public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContex
         ClusterTemplate clusterTemplateAnnot = context.getRequiredTestMethod().getDeclaredAnnotation(ClusterTemplate.class);
         if (clusterTemplateAnnot != null) {
             processClusterTemplate(context, clusterTemplateAnnot, generatedContexts::add);
-            if (generatedContexts.isEmpty()) {
-                throw new IllegalStateException("ClusterConfig generator method should provide at least one config");
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: My previous comment is wrong. This check is used to make sure there is at least one `ClusterConfig`, so we should not remove it. However, It seems to me this check should be moved to `processClusterTemplate` and we should add unit test for it.

Comment (TaiJuWu) [reply core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: If we move the check to `processClusterTemplate`, `generatedContexts` also need to expose to `processClusterTemplate`, so I prefer keep this version and revert the change and try to add unit test for it.

Comment (chia7712) [reply core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: > try to add unit test for it.

nice. It is ok to file a jira to log it and we can merge this PR first.

Comment (chia7712) [reply core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: https://issues.apache.org/jira/browse/KAFKA-16652

@TaiJuWu Please feel free to take over it

File: core/src/test/java/kafka/test/junit/ClusterTestExtensions.java [hunk 6]
```diff
@@ -118,16 +118,15 @@ public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContex
...
                                         Consumer<TestTemplateInvocationContext> testInvocations) {
         // If specified, call cluster config generated method (must be static)
         List<ClusterConfig> generatedClusterConfigs = new ArrayList<>();
-        if (!annot.value().isEmpty()) {
-            generateClusterConfigurations(context, annot.value(), generatedClusterConfigs::add);
-        } else {
-            // Ensure we have at least one cluster config
-            generatedClusterConfigs.add(ClusterConfig.defaultBuilder().build());
+
+        if (annot.value().isEmpty()) {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: How about `if (annot.value().trim().isEmpty()) throw new IllegalStateException("ClusterTemplate value can't be empty string");`

File: core/src/test/java/kafka/test/junit/ClusterTestExtensionsUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    @SuppressWarnings("unchecked")
+    void testProcessClusterTemplate() {
+        ClusterTestExtensions ext = new ClusterTestExtensions();
+        ExtensionContext context = mock(ExtensionContext.class);
+        Consumer<TestTemplateInvocationContext> testInvocations = mock(Consumer.class);
+        ClusterTemplate annot = mock(ClusterTemplate.class);
+        when(annot.value()).thenReturn("");
+
+        Assertions.assertThrows(IllegalStateException.class, () -> {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/ClusterTestExtensionsUnitTest.java]: how about `Assertions.assertThrows(IllegalStateException.class, () -> ext.processClusterTemplate(context, annot, testInvocations));`?

File: core/src/test/java/kafka/test/junit/ClusterTestExtensionsUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+public class ClusterTestExtensionsUnitTest {
+    @Test
+    @SuppressWarnings("unchecked")
+    void testProcessClusterTemplate() {
+        ClusterTestExtensions ext = new ClusterTestExtensions();
+        ExtensionContext context = mock(ExtensionContext.class);
+        Consumer<TestTemplateInvocationContext> testInvocations = mock(Consumer.class);
+        ClusterTemplate annot = mock(ClusterTemplate.class);
+        when(annot.value()).thenReturn("");
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/ClusterTestExtensionsUnitTest.java]: please add test for value "  "

**Review by chia7712 — COMMENTED** [review]
@TaiJuWu thanks for updated PR. it is almost there.

File: core/src/test/java/kafka/test/junit/ClusterTestExtensionsUnitTest.java (hunk 8 above)
Comment (TaiJuWu) [reply core/src/test/java/kafka/test/junit/ClusterTestExtensionsUnitTest.java]: Done.

**Review by chia7712 — APPROVED** [review]
LGTM

**Pull request merged**